src/gamescope/wayland/metrics/mango_app_reader.rs

```diff
@@ -0,0 +1,44 @@
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+
+use gamescope_wayland_client::mangoapp::{message::MangoAppMsgV1, queue::MangoAppMsgQueue};
+
+pub struct MangoAppMsgQueueReader {
+    queue: MangoAppMsgQueue,
+    rx: std::sync::mpsc::Receiver<()>,
+    tx: tokio::sync::watch::Sender<Option<MangoAppMsgV1>>,
+}
+
+impl MangoAppMsgQueueReader {
+    pub fn new(
+        rx: std::sync::mpsc::Receiver<()>,
+        tx: tokio::sync::watch::Sender<Option<MangoAppMsgV1>>,
+    ) -> Self {
+        log::debug!("Starting new mango app msg queue reader");
+
+        let queue = MangoAppMsgQueue::new();
+        Self { queue, rx, tx }
+    }
+
+    pub async fn run(mut self, cancel_flag: Arc<AtomicBool>) {
+        while !cancel_flag.load(Ordering::SeqCst) && self.rx.recv().is_ok() {
```

Comment:
Using a cancellation flag isn't necessary here. Using a `tokio::sync::mpsc` receiver, the receiver will return `None` when all transmitters have been dropped.

```suggestion
        while let Some(_) = self.rx.blocking_recv() {
```